Public/Src/Engine/Scheduler/Graph/PipGraph.cs

```diff
@@ -594,6 +594,22 @@ private static bool IsInput(AbsolutePath path, Pip pip)
             }
         }
 
+        /// <summary>
+        /// Returns pips consuming given directory artifact.
+        /// </summary>
+        public IEnumerable<Pip> GetConsumingPips(DirectoryArtifact dir)

```

Comment:
>GetConsumingPips [](start = 32, length = 16)

should we add some logic for composite shared opaques? E.g., 
pipA -> dir1
dir2 : compositeOf { dir1 }
pipB <- dir2